alternator/expressions.hh

```diff
@@ -26,9 +28,20 @@ public:
     using runtime_error::runtime_error;
 };
 
-parsed::update_expression parse_update_expression(std::string_view query);
-std::vector<parsed::path> parse_projection_expression(std::string_view query);
-parsed::condition_expression parse_condition_expression(std::string_view query, const char* caller);
+class expression_factory_impl;
+class expression_factory {
+    std::unique_ptr<expression_factory_impl> _impl;
+public:
+    struct config {
+        utils::updateable_value<uint32_t> max_cache_entries;
+    };
+    expression_factory(config cfg, stats& stats);
+    ~expression_factory();
+
+    parsed::update_expression parse_update_expression(std::string_view query);
+    std::vector<parsed::path> parse_projection_expression(std::string_view query);
+    parsed::condition_expression parse_condition_expression(std::string_view query, const char* caller);
```

Comment:
Seems sad to return values (that need allocations) rather than references, but don't need to improve everything at once.

Reply:
I think (but I'm not sure I'm following everything here) it anyway needs to be a writable copy, because the "references" in the expression need to be replaced by the real value.

Reply:
Well, cql3 expressions solve this by having immutable expressions evaluated by an evaluator.

We should consider merging the two, especially if we start to go crazy like jitting expressions.